src/service/navigation.js

```diff
@@ -409,6 +419,16 @@ export class Navigation {
     }
   }
 
+  /**
+   * @param {!Function} callback
+   * @param {number} priority
+   */
+  registerAnchorMutator(callback, priority) {
+    user().assert(!this.anchorMutators_[priority],
+        'Mutator with same priority is already in use.');
+    this.anchorMutators_[priority] = callback;
```

Comment:
This will make sparse arrays, which will cause `forEach` to be very slow. Use a binary-search to find the proper location, and use `Array.p.splice` to insert it instead.

Reply:
https://github.com/ampproject/amphtml/pull/17522#discussion_r210974956 😛 

I think this is fine since N is tiny, but if you really want you could use our existing `PriorityQueue` class that I wrote.

Reply:
`N` doesn't matter here, only `priority`. Given that we can't register the same priority class twice, I expect we'll see odd numbers like `100`, `200`, meaning we'll be iterating 200 items to only find 2.

Reply:
I meant `N == array.length`. 

> I expect we'll see odd numbers like 100, 200, meaning we'll be iterating 200 items to only find 2.

We can just as easily not do that by comment and example.